rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -1,226 +1,370 @@
+    # add TAG_ID_ORIGIN to the list of featurizers to make sure that we keep the
+    # tag_ids
+    featurizers.append(TAG_ID_ORIGIN)
+
+    # filter the features
+    return [f for f in features if f.origin in featurizers]
 
+
+def _create_fake_features(
+    all_features: List[List[List["Features"]]],
+) -> List["Features"]:
+    """Computes default feature values.
+
+    All given features should have the same type, e.g. dense or sparse.
+
+    Args:
+        all_features: list containing all feature values encountered in the dataset
+        for an attribute.
+
+    Returns:
+        The default features
+    """
     example_features = next(
         iter(
             [
                 list_of_features
-                for turn_features in tracker_features
-                for list_of_features in turn_features
+                for list_of_list_of_features in all_features
+                for list_of_features in list_of_list_of_features
                 if list_of_features is not None
             ]
         )
     )
 
-    # create zero_features for nones
-    zero_features = []
-    for features in example_features:
-        new_features = copy.deepcopy(features)
-        if features.is_dense():
-            new_features.features = np.zeros_like(features.features)
-        if features.is_sparse():
+    # create fake_features for Nones
+    fake_features = []
+    for _features in example_features:
+        new_features = copy.deepcopy(_features)
+        if _features.is_dense():
+            new_features.features = np.zeros(
+                (0, _features.features.shape[-1]), _features.features.dtype
+            )
+        if _features.is_sparse():
             new_features.features = scipy.sparse.coo_matrix(
-                features.features.shape, features.features.dtype
+                (0, _features.features.shape[-1]), _features.features.dtype
             )
-        zero_features.append(new_features)
+        fake_features.append(new_features)
 
-    return zero_features
+    return fake_features
 
 
 def convert_to_data_format(
-    tracker_state_features: Union[
+    features: Union[
         List[List[Dict[Text, List["Features"]]]], List[Dict[Text, List["Features"]]]
     ],
-    zero_state_features: Optional[Dict[Text, List["Features"]]] = None,
+    fake_features: Optional[Dict[Text, List["Features"]]] = None,
+    consider_dialogue_dimension: bool = True,
+    featurizers: Optional[List[Text]] = None,
 ) -> Tuple[Data, Optional[Dict[Text, List["Features"]]]]:
     """Converts the input into "Data" format.
 
+    "features_for_examples" can, for example, be a dictionary of attributes (INTENT,
+    TEXT, ACTION_NAME, ACTION_TEXT, ENTITIES, SLOTS, FORM) to a list of features for
+    all dialogue turns in all training trackers.
+    For NLU training it would just be a dictionary of attributes (either INTENT or
+    RESPONSE, TEXT, and potentially ENTITIES) to a list of features for all training
+    examples.

```

Comment:
Can we explain in the docstring what exactly is the "Data" format? It's hard to guess from just the code.

Reply:
@tabergma could you please take this over